### PR TITLE
Add function based view pylint checker #734

### PR DESCRIPTION
**Pull request by kernel-panic96**
Unfortunately I could not plug it into the tcms project, I have problems with the python virtual environment. I have created a [demo django project](https://github.com/kernel-panic96/pylint-function-based-views-demo) to showcase the plugin.

### CODE REVIEW

File: kiwi_lint/__init__.py [hunk 1]
```diff
@@ -14,6 +16,17 @@
...
+
+
+def has_django_installed():
+    """Django must be installed and DJANGO_SETTINGS_MODULE must be present in shell environment."""
+    try:
+        __import__('django')
+    except (ImportError, ModuleNotFoundError):
+        return False
+
+    return 'DJANGO_SETTINGS_MODULE' in os.environ
```
Comment (atodorov) [inline kiwi_lint/__init__.py]: Since this plugin is only valid in the current project and we are fairly certain Django is always installed we don't really need this function.

File: kiwi_lint/__init__.py [hunk 2]
```diff
@@ -28,3 +41,12 @@ def register(linter):
...
     linter.register_checker(NestedDefinitionChecker(linter))
     linter.register_checker(MissingPermissionsChecker(linter))
+
+    if has_django_installed():
+        tcms_settings = {
+            'main_urls_package': 'tcms.urls',
+            'filters': [lambda app: app.startswith('tcms')],
+        }
+        linter.register_checker(
+            FunctionBasedViewChecker(linter, **tcms_settings)
```
Comment (atodorov) [inline kiwi_lint/__init__.py]: You can pass the arguments directly to `FunctionBasedViewChecker`

File: kiwi_lint/function_based_views.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        self.url_mapping = self._prune_url_mapping(self.url_mapping, self.__installed_apps)
+        self.view_files = set(module for module, _ in self.url_mapping.values())
+
+    def _get_django_project_apps(self):
+        installed_apps = deepcopy(settings.INSTALLED_APPS)  # do not mutate the original
+
+        for _filter in self.installed_apps_function_filters:
+            # allow void filters that just mutate the list
+            res = list(_filter(installed_apps) or [])
+            installed_apps = res if res else installed_apps
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: `installed_apps = list(filter(_filter, installed_apps))` and rename `_filter` to `filter_func`

File: kiwi_lint/function_based_views.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                pruned[k] = (module, name)
+
+        return pruned
+
+    def visit_module(self, module):
+        if module.name in self.view_files:
+            if hasattr(self, 'visit_views_module'):
+                self.visit_views_module(module)
+
+    def leave_module(self, module):
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: `leave_module` and `leave_views_module` are not used and we don't seem to need them so remove both of them.

File: kiwi_lint/function_based_views.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    """
+
+    def __init__(self, *args, main_urls_package=None, filters=None, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.installed_apps_function_filters = filters or []
+        django.setup()
+
+        self.__installed_apps = self._get_django_project_apps()
+        print(self.__installed_apps)
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: Remove `print`

File: kiwi_lint/function_based_views.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    name = 'django-function-based-view-checker'
+
+    msgs = {'R4611': ('Function based view used, use class based',
+                      'non-function-based-view-required',
+                      'Function based views are the path to the Dark side!')}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.views_by_module = self.group_views_by_module(self.url_mapping.values())
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: Here I am getting wrong results:

```
['tcms.core', 'tcms.kiwi_auth', 'tcms.core.contrib.comments.apps.AppConfig', 'tcms.core.contrib.linkreference', 'tcms.management', 'tcms.testcases.apps.AppConfig', 'tcms.testplans.apps.AppConfig', 'tcms.testruns.apps.AppConfig', 'tcms.telemetry', 'tcms.xmlrpc', 'tcms.telemetry.tests.plugin']

^^^ this is from the left-over print statement above, notice the apps which are installed.

*** views by module
{'tcms.core.admin': {'delete_view', 'add_view'},
 'tcms.core.ajax': {'UpdateTestCaseActorsView',
                    'comment_case_runs',
                    'tags',
                    'update_bugs_to_caseruns'},
 'tcms.core.contrib.comments.views': {'delete', 'post'},
 'tcms.core.contrib.linkreference.views': {'add', 'remove'},
 'tcms.core.views': {'dashboard', 'navigation'},
 'tcms.kiwi_auth.admin': {'delete_view', 'user_change_password'},
 'tcms.kiwi_auth.views': {'LoginViewWithCustomTemplate',
                          'PasswordResetView',
                          'confirm',
                          'profile',
                          'register'},
 'tcms.telemetry.tests.plugin.views': {'example'},
 'tcms.telemetry.views': {'TestingBreakdownView'}}
```

^^^ this is from a print I've added here. Notice how you have both class based and function based views in the list and also how you are missing every app which is declared not by its module name but via its AppConfig.

Comment (kernel-panic96) [reply kiwi_lint/function_based_views.py]: That was my original intention the DjangoViewsVisiter class only gathers views which are directly or transitively included in the project's `urlpatterns`, the idea was that by doing it so you could find views which are not listed in a standard 'views.py' file, e.g. tcms.kiwi_auth.admin's views. If your idea was different, I'll be happy to change the implementation.

If you don't see an module in the self.views_by_module (I'm assuming that this was what you printed out), then the module did not contain views who are routed in the project's urlpatterns.  Can you elaborate on the last AppConfig, are you reffering to [this](https://docs.djangoproject.com/en/2.2/ref/applications/#for-application-users)?

Comment (kernel-panic96) [reply kiwi_lint/function_based_views.py]: I committed a change, that should be able to deal with AppConfig's. Could you please verify the behaviour on the project, I still have issues running it.

File: kiwi_lint/function_based_views.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+from django.urls.resolvers import URLResolver, URLPattern
+
+from pylint import checkers
+from pylint import interfaces
+
+
+class DjangoViewsVisiter(checkers.BaseChecker):
+    """Base class for visiting only astroid modules which contain django views.
+
+    Derivatives could override `visit_views_module` and/or `leave_views_module` to hook just into
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: `leave_views_module` is no more.

File: kiwi_lint/function_based_views.py (hunk 6 above)
Comment (atodorov) [reply kiwi_lint/function_based_views.py]: With the latest update I have the following views by module:

```
{'tcms.core.admin': {'add_view', 'delete_view'},
 'tcms.core.ajax': {'UpdateTestCaseActorsView',
                    'comment_case_runs',
                    'tags',
                    'update_bugs_to_caseruns'},
 'tcms.core.contrib.comments.views': {'delete', 'post'},
 'tcms.core.contrib.linkreference.views': {'add', 'remove'},
 'tcms.core.views': {'navigation', 'dashboard'},
 'tcms.kiwi_auth.admin': {'user_change_password', 'delete_view'},
 'tcms.kiwi_auth.views': {'LoginViewWithCustomTemplate',
                          'PasswordResetView',
                          'confirm',
                          'profile',
                          'register'},
 'tcms.telemetry.tests.plugin.views': {'example'},
 'tcms.telemetry.views': {'TestingBreakdownView'},
 'tcms.testcases.admin': {'add_view', 'change_view'},
 'tcms.testcases.views': {'NewCaseView',
                          'SimpleTestCaseView',
                          'TestCaseExecutionDetailPanelView',
                          'attachment',
                          'clone',
                          'edit',
                          'get',
                          'list_all',
                          'load_more_cases',
                          'printable',
                          'search'},
 'tcms.testplans.admin': {'add_view', 'change_view'},
 'tcms.testplans.views': {'DeleteCasesView',
                          'LinkCasesSearchView',
                          'LinkCasesView',
                          'NewTestPlanView',
                          'ReorderCasesView',
                          'UpdateParentView',
                          'attachment',
                          'clone',
                          'edit',
                          'get',
                          'get_all',
                          'printable',
                          'search'},
 'tcms.testruns.admin': {'add_view', 'change_view', 'delete_view'},
 'tcms.testruns.views': {'AddCasesToRunView',
                         'TestRunReportView',
                         'UpdateAssigneeView',
                         'UpdateCaseRunStatusView',
                         'cc',
                         'change_status',
                         'clone',
                         'edit',
                         'get',
                         'new',
                         'remove_execution',
                         'search',
                         'update_case_run_text'}}
```

Issues:
- `admin.py` is not included directly in urlpatterns and it doesn't define any views so being present is wrong.
- How did `'tcms.telemetry.tests.plugin.views': {'example'},` get included? It is not referenced from `tcms/urls.py` ?

File: kiwi_lint/function_based_views.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    @staticmethod
+    def group_views_by_module(module_view_name_pairs):
+        by_module = {}
+        for module, identifier in module_view_name_pairs:
+            by_module.setdefault(module, set()).add(identifier)
+
+        return by_module
+
+    def visit_views_module(self, module):
+        for func_def in module.nodes_of_class(astroid.FunctionDef):
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: You can simplify this by just having `visit_module` to make sure you are in a module which has views (will set a boolean flag) and `visit_functiondef` to examine if this is a FBV and call `add_message`.

File: kiwi_lint/function_based_views.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    def __init__(self, *args, main_urls_package=None, filters=None, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.installed_apps_function_filters = filters or []
+        django.setup()
+
+        self.__installed_apps = self._resolve_app_configs(self._get_django_project_apps())
+
+        project_urls = import_module(main_urls_package)
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: This is always `settings.ROOT_URLCONF` (points to the urls module) so you don't need to pass this externally as parameter.

File: kiwi_lint/__init__.py [hunk 10]
```diff
@@ -34,3 +35,12 @@ def register(linter):
     linter.register_checker(NestedDefinitionChecker(linter))
     linter.register_checker(MissingPermissionsChecker(linter))
     linter.register_checker(AutoFieldChecker(linter))
+    linter.register_checker(
+        FunctionBasedViewChecker(
+            linter,
+            main_urls_package='tcms.urls',
+            filters=[
+                lambda apps: filter(lambda app: app.startswith('tcms'), apps)
```
Comment (atodorov) [inline kiwi_lint/__init__.py]: Why exactly do we need to have filters ? I tried without it and everything seems to work just fine.

From what I can tell the filters will only limit the list of modules which the checker collects but that has nothing to do with the list of python files which pylint inspects. That is if I don't specify filters I get several extra view modules in the data structure (e.g. modernrpc.views) but because pylint is inspecting only files in the tcms/ directory I never get to do a lookup for these keys in the dictionary.

IMO you can remove everything related to filters and simplify even more.

Comment (kernel-panic96) [reply kiwi_lint/__init__.py]: Good call, I did not take into account that you invoke pylint on specific files.

File: Makefile [hunk 11]
```diff
@@ -52,7 +52,7 @@ check: flake8 test
 .PHONY: pylint
 pylint:
 	pylint -d missing-docstring *.py kiwi_lint/
-	PYTHONPATH=.:./tcms/ pylint --load-plugins=pylint_django --load-plugins=kiwi_lint -d missing-docstring -d duplicate-code tcms/
+	PYTHONPATH=.:./tcms/ DJANGO_SETTINGS_MODULE=$(DJANGO_SETTINGS_MODULE) pylint --load-plugins=pylint_django --load-plugins=kiwi_lint -d missing-docstring -d duplicate-code tcms/
```
Comment (atodorov) [inline Makefile]: Revert this change and add `-d non-function-based-view-required` to it!

Then add the following line:

```
 PYTHONPATH=. DJANGO_SETTINGS_MODULE=$(DJANGO_SETTINGS_MODULE) pylint --load-plugins=pylint_django --load-plugins=kiwi_lint -d all -e  non-function-based-view-required tcms/
```

Effectively we want pylint to execute twice, once without the new checker and once with the new checker enabled. This will resolve the import problem with the `xmlrpc` module.

File: kiwi_lint/function_based_views.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        django.setup()
+
+        project_urls = import_module(settings.ROOT_URLCONF)
+        self.url_mapping = self._get_url_view_mapping(project_urls.urlpatterns)
+        self.view_files = set(module for module, _ in self.url_mapping.values())
+        self.view_module = None
+
+    @classmethod
+    def _get_url_view_mapping(cls, root_urlpatterns):
+        def go(urlpatterns, prefix='^', result=None):
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: Name must be longer and more descriptive. Both CodeClimate and pylint complain about it.

File: kiwi_lint/function_based_views.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+                elif isinstance(url, URLResolver):
+                    # path('someurl', include(some_url_patterns)), recurse on some_url_patterns
+                    go(url.url_patterns, prefix + url.pattern.regex.pattern.strip('^$'), result)
+
+            return result
+
+        return go(root_urlpatterns)
+
+    def visit_module(self, module):
+        self.view_module = module.name if module.name in self.view_files else None
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: Skip modules ending in `.admin`, ether here or when scanning `urlpatterns`.

File: kiwi_lint/function_based_views.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+            """
+
+            if result is None:
+                result = {}
+
+            for url in urlpatterns:
+                if isinstance(url, URLPattern):
+                    # path('someurl', view), meaning this is leaf node url
+                    url_pattern = prefix + url.pattern.regex.pattern.strip('^')
+                    result[url_pattern] = (url.callback.__module__, url.callback.__name__)
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: `result[rl.callback.__module__] = url.callback.__name__` will directly provide the data structure expected by the checker and negate the need for `group_views_by_module()` below.

File: kiwi_lint/function_based_views.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        by_module = {}
+        for module, identifier in module_view_name_pairs:
+            by_module.setdefault(module, set()).add(identifier)
+
+        return by_module
+
+    def visit_functiondef(self, func_def):
+        if self.view_module and func_def.name in self.views_by_module[self.view_module]:
+            self.add_message('non-function-based-view-required', node=func_def)
+            print(f'\tview name: \'{func_def.name}\'')
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: Remove print.

File: kiwi_lint/function_based_views.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    """
+
+    def __init__(self, linter):
+        super().__init__(linter)
+
+        django.setup()
+
+        project_urls = import_module(settings.ROOT_URLCONF)
+        self.url_mapping = self._get_url_view_mapping(project_urls.urlpatterns)
+        self.view_files = set(module for module, _ in self.url_mapping.values())
```
Comment (atodorov) [inline kiwi_lint/function_based_views.py]: with the proposed refactoring of the internal data structure you don't really need this variable, just take `self.url_mapping.keys()`
